### PR TITLE
ENH: add support for fujitsu compiler to numpy.

### PR DESCRIPTION
**Pull request by t-karatsu**
Hello, this is my first post.
If I need to do something else about submitting PR, please let me know.

I would like to add support for Fujitsu compiler to numpy, so submitted the necessary changes.
Could you check if there are any questions or areas that need to be corrected?

### CODE REVIEW

**Comment by rgommers** [comment]
Isn't this the  Lahey/Fujits compiler already supported in `fcompiler/lahey.py`?

**Comment by t-karatsu** [comment]
@rgommers 
Thanks for your reply.

> Isn't this the Lahey/Fujitsu compiler already supported in fcompiler/lahey.py?

No, it's different from the lahey/fujitsu compiler.
The new compiler registered is developed by Fujitsu and is used in A64FX on Fugaku. There is currently no HP that introduces this compiler.

File: numpy/distutils/fcompiler/__init__.py [hunk 1]
```diff
@@ -744,7 +744,7 @@ def wrap_unlinkable_objects(self, objects, output_dir, extra_dll_dir):
                'intelvem', 'intelem', 'flang')),
     ('cygwin.*', ('gnu', 'intelv', 'absoft', 'compaqv', 'intelev', 'gnu95', 'g95')),
     ('linux.*', ('gnu95', 'intel', 'lahey', 'pg', 'nv', 'absoft', 'nag', 'vast', 'compaq',
-                 'intele', 'intelem', 'gnu', 'g95', 'pathf95', 'nagfor')),
+                 'intele', 'intelem', 'gnu', 'g95', 'pathf95', 'nagfor', 'fj')),
```
Comment (rgommers) [inline numpy/distutils/fcompiler/__init__.py:747]: Can you call it either `frt` (after the executable) or `fujitsu`? Just `fj` is not descriptive enough.

**Review by rgommers — COMMENTED** [review]
> No, it's different from the lahey/fujitsu compiler.
> The new compiler registered is developed by Fujitsu and is used in A64FX on Fugaku. 

Thanks, that helps. Could you add this info to the docstring of the file? Including a link to the current best source of information about it?

> There is currently no HP that introduces this compiler.

That doesn't preclude inclusion here I think; there's no other good way to add compiler support and it's very low-overhead so I'm fine with merging this.

One question out of interest - if `frt` going to stay? Based on [this article](https://www.phoronix.com/scan.php?page=news_item&px=Fujitsu-A64FX-GCC-Start) I'd expect `gfortran` support too, which may then become the default?

File: numpy/distutils/fcompiler/__init__.py (hunk 1 above)
Comment (t-karatsu) [reply numpy/distutils/fcompiler/__init__.py]: Thank you for confirmation. From now on, the name of the compiler type will be fixed at fujitsu.

**Comment by t-karatsu** [comment]
> Could you add this info to the docstring of the file?

OK, I will add docstring to 'fujitsu.py'.
I'm planning to publish a following article introducing a64fx:
https://www.fujitsu.com/global/products/computing/servers/supercomputer/a64fx/

> One question out of interest - if frt going to stay? Based on this article I'd expect gfortran support too, which may then become the default?

As you say, we can use the gcc compiler on Fugaku, but we recommend using the Fujitsu compiler to maximize the performance of fugaku.

I just agree to support fugaku's gfortran in numpy.
I can't determine if a special implementation about gfortran is needed, so I'll ask the person in charge to consider gfortran support.

**Comment by rgommers** [comment]
> As you say, we can use the gcc compiler on Fugaku, but we recommend using the Fujitsu compiler to maximize the performance of fugaku.

That makes sense.

> I just agree to support fugaku's gfortran in numpy.

Sounds good, thank you.

**Review by rgommers — APPROVED** [review]
LGTM now, merging. Thanks @t-karatsu

**Comment by t-karatsu** [comment]
Thanks for merge!

**Comment by migueldiascosta** [comment]
@t-karatsu this PR added compiler support but performance is still poor because the fujitsu linear algebra library is not detected. 

Of course, one can build e.g. openblas with the fujitsu compiler and it get's better, but it would be better to use the fujitsu linear algebra library, no? As I understand it, it's always (?) bundled with the compiler?

The patch in https://gist.github.com/migueldiascosta/3bf319b9eecd888957db50933d2d701f worked for me at Fugaku, and performance is indeed better - is this something you've worked on and/or are interested in testing/reviewing?

**Pull request merged**